state/state.go

```diff
@@ -1,29 +1,68 @@
 // Copyright © 2019 Arrikto Inc.  All Rights Reserved.
 
-package main
+package state
 
 import (
-	"github.com/gorilla/sessions"
-	"github.com/pkg/errors"
+	"math/rand"
 	"net/http"
 	"net/http/httptest"
 	"time"
+
+	"github.com/gorilla/sessions"
+	"github.com/pkg/errors"
 )
 
 const oidcLoginSessionCookie = "non_existent_cookie"
 
+var nonceChars = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
+
 type state struct {
-	origURL string
+	OrigURL string
 }
 
-func newState(origURL string) *state {
+type Config struct {
+	SchemeDefault string
+	SchemeHeader  string
+	SchemeAndHost bool
+}
+
+type StateFunc func(*http.Request) *state
+
+func NewStateFunc(config *Config) StateFunc {
+	if config.SchemeAndHost {
+		return newSchemeAndHost(config)
+	}
+	return relativeURL
+}
+
+func relativeURL(r *http.Request) *state {
```

Comment:
I'd just remove this entirely.

Reply:
Would like @yanniszark input here. I was trying to preserve the original functionality by keeping the relativeURL method in absence of a user specified SESSION_DOMAIN.